### PR TITLE
Bromley waste phase2 payments

### PR DESCRIPTION
**Pull request by struan**
[skip changelog]

### CODE REVIEW

File: t/app/controller/waste.t [hunk 1]
```diff
@@ -257,14 +380,709 @@ FixMyStreet::override_config {
...
+            { order_by => { -desc => 'id' } },
+        )->first;
+
+        is $new_report->category, 'Cancel Garden Subscription', 'correct category on report';
+        is $new_report->get_extra_field_value('Container_Request_Details_Action'), 2, 'correct container request action';
+        is $new_report->get_extra_field_value('Container_Request_Details_Quantity'), 1, 'correct container request count';
+        is $new_report->state, 'confirmed', 'report confirmed';
+    };
+
+    subtest 'check new sub credit card payment lookup failed' => sub {
```
Comment (dracos) [inline t/app/controller/waste.t:943]: This and next test seem identical apart from the mock - refactor? Also, they're testing what the code does but I'm not sure that's a great outcome for the user at present

File: perllib/FixMyStreet/Cobrand/Bromley.pm [hunk 2]
```diff
@@ -968,6 +1049,61 @@ sub waste_check_last_update {
...
+
+    return $now > $sub_end;
+}
+
+sub garden_waste_cost {
+    my ($self, $bin_count) = @_;
+
+    $bin_count ||= 1;
+
+    return $self->feature('payment_gateway')->{ggw_cost} * $bin_count;
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Bromley.pm:1104]: There are a number of places that make this calculation and don't use this function.

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 3]
```diff
@@ -562,16 +719,134 @@ sub process_garden_data : Private {
     $data->{title} = $data->{category};
     $data->{detail} = "$data->{category}\n\n$address";
 
-    $c->set_param('service_id', $data->{service_id});
-    $c->set_param('Subscription_Type', $c->stash->{garden_subs}->{New});
+    $c->set_param('service_id', $c->stash->{services_available}->{Garden_Waste}->{service_id});
```
Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Waste.pm:722]: There's the special garden ID function?

File: perllib/FixMyStreet/App/Form/Waste/Garden/Renew.pm [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+has_field billing_differ => (
+    type => 'Checkbox',
+    option_label => 'Check if different to collection address',
+    label => "Billing address",
+    tags => {
+        toggle => 'form-billing_address-row'
+    },
+);
+
+has_field billing_address => (
```
Comment (dracos) [inline perllib/FixMyStreet/App/Form/Waste/Garden/Renew.pm:103]: Can this share code with the garden subscription form, if all these fields are the same?

**Review by dracos — CHANGES_REQUESTED** [review]
This is a big PR, which has made it hard (and long) to look through, perhaps e.g. the caching could have been a separate PR, the renewal form bit, etc? (I realise they were separate commits, which definitely helped and I looked through them individually but e.g. waste_sub_due was used in a commit before it was defined.) I'm not sure I have managed to check or have caught everything, but I'm afraid there's quite a lot of comments nevertheless.

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 5]
```diff
@@ -622,7 +646,36 @@ sub garden_renew : Chained('property') : Args(0) {
...
+    my $payment_method = $c->stash->{orig_sub} ?
+        $c->stash->{orig_sub}->get_extra_field_value('payment_method') :
+        'credit_card';
+
+    if ( FixMyStreet->staging_flag('skip_waste_payment') ) {
+        $c->stash->{report}->confirm;
+        $c->stash->{report}->update;
+    } else {
+        if ( $payment_method eq 'direct_debit' ) {
+            $c->forward('direct_debit_cancel_sub');
```
[the remark this answers is not in the document]
Comment (struan) [reply perllib/FixMyStreet/App/Controller/Waste.pm:682]: All the direct debit processing takes place in a separate process which I've not written yet which will do the confirming etc.

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 6]
```diff
@@ -254,6 +261,17 @@ sub direct_debit_modify : Path('dd_amend') : Args(0) {
...
 
+sub direct_debit_cancel_sub : Path('dd_cancel_sub') : Args(0) {
+    my ($self, $c) = @_;
+
+    my $p = $c->stash->{report};
+
+    my $i = Integrations::Pay360->new( { config => $c->cobrand->feature('payment_gateway') } );
+
+    my $update_ref = $i->cancel_plan( {
+        payer_reference => 1, # XXX
```
[the remark this answers is not in the document]
Comment (struan) [reply perllib/FixMyStreet/App/Controller/Waste.pm:281]: That's another thing that's being set up in the DD processing so it's awaiting that code before I know what's going on with it.

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 7]
```diff
@@ -123,10 +135,10 @@ sub pay : Path('pay') : Args(0) {
 sub pay_complete : Path('pay_complete') : Args(1) {
```
[the remark this answers is not in the document]
Comment (struan) [reply perllib/FixMyStreet/App/Controller/Waste.pm:161]: There isn't a hook option with the card payment API we're using.

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 8]
```diff
@@ -547,15 +587,29 @@ sub process_garden_cancellation : Private {
...
-sub process_garden_modification : Private {
-    my ($self, $c, $form) = @_;
-    # TODO Modify subscription here
+sub get_original_sub : Private {
+    my ($self, $c) = @_;
+
+    my $p = $c->model('DB::Problem')->search({
+        user_id => $c->user->id,
+        category => 'New Garden Subscription',
+        extra => { like => '%property_id,T5:value,I_:'. $c->stash->{property}{id} . '%' }
```
[the remark this answers is not in the document]
Comment (struan) [reply perllib/FixMyStreet/App/Controller/Waste.pm:734]: I think the state is going to come from Echo as we're going to store the payment method in there so this code needs to change to reflect that. However I'm waiting on Echo being updated with that at the moment.

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 9]
```diff
@@ -147,25 +159,36 @@ sub pay_complete : Path('pay_complete') : Args(1) {
             # cancelled, not attempted, logged out - try again option
             # card rejected - try again with different card/cancel
             # otherwise error page?
+            $c->stash->{template} = 'waste/pay.html';
             $c->stash->{error} = 'Payment failed: ' . $resp->{paymentResult}->{status};
+            $c->detach;
         }
     } else {
         # retry if in progress, error if invalid ref.
```
[the remark this answers is not in the document]
Comment (struan) [reply perllib/FixMyStreet/App/Controller/Waste.pm:195]: I'm going to fix this in the branch that fixes the card payments

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 10]
```diff
@@ -147,25 +159,36 @@ sub pay_complete : Path('pay_complete') : Args(1) {
             # cancelled, not attempted, logged out - try again option
```
[the remark this answers is not in the document]
Comment (struan) [reply perllib/FixMyStreet/App/Controller/Waste.pm:187]: I'm going to fix this in the branch that fixes the card payments

**Review by dracos — CHANGES_REQUESTED** [review]
Great, looks like everything I raised has been sorted. Only issue is I suggest the return URL is pay_complete/ID/random_token which lets you look it up quickly rather than scan, and is more random.

File: t/app/controller/waste.t [hunk 11]
```diff
@@ -474,16 +480,23 @@ FixMyStreet::override_config {
...
 
         is $new_report->category, 'New Garden Subscription', 'correct category on report';
         is $new_report->get_extra_field_value('payment_method'), 'direct_debit', 'correct payment method on report';
         is $new_report->state, 'unconfirmed', 'report not confirmed';
 
-        $mech->get_ok("/waste/dd_complete?reference=$report_id");
+        $mech->get("/waste/dd_complete?reference=$token&report_id=xxy");
+        ok !$mech->res->is_success(), "want a bad response";
+        is $mech->res->code, 404, "got 404";
+        $mech->get("/waste/dd_complete?reference=NOTATOKENreport_id=$report_id");
```
Comment (dracos) [inline t/app/controller/waste.t:496]: Should there be an `&` here?

**Review by dracos — APPROVED** [review]
One typo in the test I think is all :)

**Pull request merged**